# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/workers/WorkerDebuggerManager.cpp
# Commit: 1e774bd7f365
# Full Hash: 1e774bd7f36525a03b6260b1cd7741bcf4e5f622
# Author: Eden Chuang <echuang@mozilla.com>
# Date: 2025-04-09 04:12:49
# Regressor Bug: 1899503
# File Overlap Count: 4
# Description:
#   Bug 1899503 - P3 PRemoteWorkerDebugger binding implementation. r=asuth
#   
#   This patch implements the PRemoteWorkerDebugger binding.
#   
#   We are trying to keep the same logic debugger registration/unregistration to minimize the difference between the local debugger(the debugger on the content process main thread)
# ==============================================================================

diff -r b003e26209cf -r 1e774bd7f365 dom/workers/WorkerDebuggerManager.cpp
--- a/dom/workers/WorkerDebuggerManager.cpp	Tue Apr 08 09:43:33 2025 +0000
+++ b/dom/workers/WorkerDebuggerManager.cpp	Tue Apr 08 09:43:33 2025 +0000
@@ -77,7 +77,15 @@
  public:
   explicit WorkerDebuggerEnumerator(
       const nsTArray<nsCOMPtr<nsIWorkerDebugger>>& aDebuggers)
-      : mDebuggers(aDebuggers.Clone()), mIndex(0) {}
+      : mIndex(0) {
+    for (auto debugger : aDebuggers) {
+      bool isRemote;
+      Unused << debugger->GetIsRemote(&isRemote);
+      if (!isRemote) {
+        mDebuggers.AppendElement(debugger);
+      }
+    }
+  }
 
   NS_DECL_NSISIMPLEENUMERATOR
 
@@ -275,10 +283,9 @@
   AssertIsOnMainThread();
 
   mDebuggers.AppendElement(aRemoteWorkerDebugger);
-
-  for (const auto& listener : CloneListeners()) {
-    listener->OnRegister(aRemoteWorkerDebugger);
-  }
+  //  for (const auto& listener : CloneListeners()) {
+  //    listener->OnRegister(aRemoteWorkerDebugger);
+  //  }
 }
 
 void WorkerDebuggerManager::UnregisterDebugger(
@@ -287,10 +294,9 @@
   AssertIsOnMainThread();
 
   mDebuggers.RemoveElement(aRemoteWorkerDebugger);
-
-  for (const auto& listener : CloneListeners()) {
-    listener->OnUnregister(aRemoteWorkerDebugger);
-  }
+  //  for (const auto& listener : CloneListeners()) {
+  //    listener->OnUnregister(aRemoteWorkerDebugger);
+  //  }
 }
 
 void WorkerDebuggerManager::RegisterDebuggerMainThread(
@@ -349,8 +355,10 @@
   MOZ_ASSERT_DEBUG_OR_FUZZING(!aWorkerId.IsEmpty());
   for (auto debugger : mDebuggers) {
     nsAutoString workerId;
+    bool isRemote;
     debugger->GetId(workerId);
-    if (workerId.Equals(aWorkerId)) {
+    debugger->GetIsRemote(&isRemote);
+    if (workerId.Equals(aWorkerId) && isRemote) {
       return debugger;
     }
   }